fiat-roles/src/main/java/com/netflix/spinnaker/fiat/roles/UserRolesSyncStrategy.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2022 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.fiat.roles;
+
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@FunctionalInterface
+public interface UserRolesSyncStrategy {
+  long syncAndReturn(List<String> roles);
+
+  @RequiredArgsConstructor
+  class DefaultSynchronizationStrategy implements UserRolesSyncStrategy {
+
+    private final Synchronizer synchronizer;
+
+    @Override
+    public long syncAndReturn(List<String> roles) {
+      return this.synchronizer.syncAndReturn(roles);
+    }
+  }
+
+  @Slf4j
+  class CachedSynchronizationStrategy implements UserRolesSyncStrategy {
+
+    private final Synchronizer synchronizer;
+    private final CallableCache<List<String>, Long> callableCache;
+
+    public CachedSynchronizationStrategy(Synchronizer synchronizer) {
+      this.synchronizer = synchronizer;
+      this.callableCache = new CallableCache<>();
+    }
+
+    public CachedSynchronizationStrategy(
+        Synchronizer synchronizer, CallableCache<List<String>, Long> cache) {
+      this.synchronizer = synchronizer;
```

Comment:
Not sure if it would make sense to validate that the constructor parameters are not null? Just to make sure that whoever is instantiating this class is not passing a null param by mistake?

Reply:
@ovidiupopa07 I used lombok's annotation @NonNull. Is that ok?

Reply:
yep, sure 👍